### PR TITLE
BZ 1299000 -- Document uploading images to the internal registry

### PR DESCRIPTION
**Pull request by tnguyen-rh**
This is for [BZ 1299000](https://bugzilla.redhat.com/show_bug.cgi?id=1299000).
It adds an xref from the Creating Images "Overview" to Install/Config "Docker Registry", highlights the term `pull specification` in Architecture, emphasizes the `<project>` component of the pull specification in the `docker tag` command, and adds a section on the registry's IP address (pulling from https://github.com/openshift/openshift-docs/issues/1494).

### CODE REVIEW

**Comment by tnguyen-rh** [comment]
@pweil- @bfallonf 
PTAL

File: install_config/install/docker_registry.adoc [hunk 1]
```diff
@@ -169,9 +169,41 @@ $ oadm registry --service-account=registry \
...
+[[registry-ip-address]]
+=== Registry IP Address
+
+Because the integrated registry is referenced by its service IP address,
+if you decide to delete and recreate the docker-registry service,
+you should either take care to re-use its address,
+or reboot the cluster after creating the new docker-registry service.
+
+[[re-using-the-address]]
+==== Re-using the Address
```
Comment (bfallonf) [inline install_config/install/docker_registry.adoc]: @tnguyen-rh  Did we ever talk about never going more than 3 ==='s down? There's a chance I made that rule up in my head...

I feel you could have the same effect here by calling "Re-using the Address" and "Rebooting the Cluster" into " *Option 1: Re-using the Address* " and the like.

File: install_config/install/docker_registry.adoc [hunk 2]
```diff
@@ -169,9 +169,41 @@ $ oadm registry --service-account=registry \
...
+service prior to deleting it, and arrange to replace the newly assigned IP address
+with the saved one in the new docker-registry service.
+
+// NB: Snarfed from <https://github.com/openshift/openshift-docs/issues/1494>.
+. Make a note of the `ClusterIP` for the service.
+. `oc delete svc/docker-registry dc/docker-registry`
+. `oadm registry ... -o yaml > registry.yaml`
+. Edit *_registry.yaml_*, find the `Service` there,
+and set its `ClusterIP` to the address noted in step 1.
+. `oc create -f registry.yaml`
```
Comment (bfallonf) [inline install_config/install/docker_registry.adoc]: The above doesn't seem to mark up correctly. Too many periods before each command?

File: install_config/install/docker_registry.adoc [hunk 3]
```diff
@@ -169,9 +169,41 @@ $ oadm registry --service-account=registry \
...
+you must reboot the cluster after creating the new docker-registry service.
+In some circumstances it is sufficient to reboot only the master,
+but that may leave stale references to the old IP address in
+link:../../architecture/infrastructure_components/kubernetes_infrastructure.html#master[etcd].
+
+
+[[viewing-logs]]
 == Viewing Logs
 
 To view the logs for the Docker registry, run the `oc logs` indicating the desired pod:
```
Comment (bfallonf) [inline install_config/install/docker_registry.adoc]: I know this was here before this PR, but "run the `oc logs`" might need a "command" thrown in there, or a deleting of "the"

"run `oc logs`"
"run the `oc logs` command"

**Comment by bfallonf** [comment]
@tnguyen-rh A few pokes :point_right:

File: install_config/install/docker_registry.adoc (hunk 1 above)
Comment (tnguyen-rh) [inline install_config/install/docker_registry.adoc]: Yeah, i had a feeling this was too deep (and the volume in each sub-sub-sub section too little to warrant it).  Will rework to use a labelled list.

File: install_config/install/docker_registry.adoc (hunk 2 above)
Comment (tnguyen-rh) [inline install_config/install/docker_registry.adoc]: Good idea.  Will do.

File: install_config/install/docker_registry.adoc (hunk 3 above)
Comment (tnguyen-rh) [inline install_config/install/docker_registry.adoc]: I suppose even better would be to omit the inline markup.  But, as you note, this is out of scope for this PR.  @vikram-redhat WDYT?

Comment (kargakis) [inline install_config/install/docker_registry.adoc]: We know support getting application logs just by running oc logs against the dc

Comment (kargakis) [inline install_config/install/docker_registry.adoc]: So once the docker registry is deployed, `oc logs dc/docker-registry` should be the same as `oc logs docker-registry-1-da73t`.

File: install_config/install/docker_registry.adoc [hunk 4]
```diff
@@ -169,9 +169,41 @@ $ oadm registry --service-account=registry \
...
 
-[[viewing-logs]]
 
+[[registry-ip-address]]
+=== Registry IP Address
+
+Because the integrated registry is referenced by its service IP address,
+if you decide to delete and recreate the docker-registry service,
+you should either take care to re-use its address,
+or reboot the cluster after creating the new docker-registry service.
```
Comment (ncdc) [inline install_config/install/docker_registry.adoc]: As mentioned in the bz, the only thing you need to restart is the master api. Restarting anything else will cause unnecessary downtime.

File: install_config/install/docker_registry.adoc [hunk 5]
```diff
@@ -169,9 +169,41 @@ $ oadm registry --service-account=registry \
...
+. `oc create -f registry.yaml`
 
+[[rebooting-the-cluster]]
+Rebooting the Cluster::
+
+If you are unable to re-use the IP address,
+you must reboot the cluster after creating the new docker-registry service.
+In some circumstances it is sufficient to reboot only the master,
+but that may leave stale references to the old IP address in
+link:../../architecture/infrastructure_components/kubernetes_infrastructure.html#master[etcd].
```
Comment (ncdc) [inline install_config/install/docker_registry.adoc]: Same comment as above. Only restart the master api.

File: install_config/install/docker_registry.adoc [hunk 6]
```diff
@@ -398,7 +430,11 @@ $ docker pull docker.io/busybox
 ----
 ====
 
-. Tag the new image with the form `<registry_ip>:<port>/<project>/<image>`:
+. Tag the new image with the form `<registry_ip>:<port>/<project>/<image>`.
```
Comment (ncdc) [inline install_config/install/docker_registry.adoc]: You can also append a tag after `<image>`. If you omit it, it defaults to latest.

Comment (tnguyen-rh) [inline install_config/install/docker_registry.adoc]: Right.  I'll mention that in the intro blurb.

File: install_config/install/docker_registry.adoc (hunk 3 above)
Comment (tnguyen-rh) [inline install_config/install/docker_registry.adoc]: Thanks for the tip.  I think this definitely warrants another issue:
https://github.com/openshift/openshift-docs/issues/1645

**Comment by tnguyen-rh** [comment]
@bfallonf 
Thanks for the review.  With the exception of the `oc log`-related change (which has its [own issue](#1645) now), i've installed the suggested changes.
@ncdc 
I've bifurcated the "reboot cluster" section to describe "only master" and "entire cluster" scenarios.  The reason i want to keep "entire cluster" is abstracted to *NORMALLY-HARMLESS* and *MILDLY-UNDESIRABLE-CONSEQUENCE*.  Could you suggest some text for those placeholders?

File: install_config/install/docker_registry.adoc [hunk 7]
```diff
@@ -169,9 +170,68 @@ $ oadm registry --service-account=registry \
...
 
-[[viewing-logs]]
 
+[[registry-ip-address]]
+=== Registry IP Address
+
+Because the integrated registry is referenced by its service IP address,
+if you decide to delete and recreate the docker-registry service,
+you should either take care to re-use its address,
+or reboot some or all of the cluster after creating the new docker-registry service.
```
Comment (ncdc) [inline install_config/install/docker_registry.adoc]: Please revise this to only mention restarting the masters.

File: install_config/install/docker_registry.adoc [hunk 8]
```diff
@@ -169,9 +170,68 @@ $ oadm registry --service-account=registry \
...
++
+----
+oc create -f registry.yaml
+----
+
+[[rebooting-the-cluster]]
+Rebooting the Cluster::
+
+If you are unable to re-use the IP address,
+you must reboot some or all of the cluster after creating the new docker-registry service.
```
Comment (ncdc) [inline install_config/install/docker_registry.adoc]: Please revise this to only mention restarting the masters.

File: install_config/install/docker_registry.adoc [hunk 9]
```diff
@@ -169,9 +170,68 @@ $ oadm registry --service-account=registry \
...
+is cleared from the cache.
++
+[horizontal]
+Only master:::
+Rebooting only the master avoids unnecessary downtime.
+However, it may leave stale references to the old IP address in
+link:../../architecture/infrastructure_components/kubernetes_infrastructure.html#master[etcd].
+This is *NORMALLY-HARMLESS* but *MILDLY-UNDESIRABLE-CONSEQUENCE*.
+
+Entire cluster:::
```
Comment (ncdc) [inline install_config/install/docker_registry.adoc]: Please remove this section

File: install_config/install/docker_registry.adoc [hunk 10]
```diff
@@ -169,9 +170,68 @@ $ oadm registry --service-account=registry \
...
+
+If you are unable to re-use the IP address,
+you must reboot some or all of the cluster after creating the new docker-registry service.
+This ensures that the old registry URL, which includes the old IP address,
+is cleared from the cache.
++
+[horizontal]
+Only master:::
+Rebooting only the master avoids unnecessary downtime.
+However, it may leave stale references to the old IP address in
```
Comment (ncdc) [inline install_config/install/docker_registry.adoc]: There is no way around stale IP addresses if the docker-registry service's IP changes. You either have to preserve the original IP, or write some utility/script to traverse etcd and find-and-replace the old IP with the new IP.

**Comment by ncdc** [comment]
@tnguyen-rh 

> I've bifurcated the "reboot cluster" section to describe "only master" and "entire cluster" scenarios. The reason i want to keep "entire cluster" is abstracted to NORMALLY-HARMLESS and MILDLY-UNDESIRABLE-CONSEQUENCE. Could you suggest some text for those placeholders?

I've added comments to individual lines. In summary, please remove any reference to rebooting the entire cluster.

cc @pweil- @miminar

File: install_config/install/docker_registry.adoc (hunk 10 above)
Comment (tnguyen-rh) [inline install_config/install/docker_registry.adoc]: (from IRC)
> [R]ebooting the cluster does not remove stale registry IPs.
Thanks for the clarification.  This was the key misunderstanding.

File: install_config/install/docker_registry.adoc (hunk 9 above)
Comment (tnguyen-rh) [inline install_config/install/docker_registry.adoc]: (from IRC in response to followup question "what is the consequence of stale IP addrs?")

> possibility of failed image pulls (leading to failed container/pod starts) and pushes

and also

> if you restart the master, so it picks up the new ip, new builds & deployments will start using the new ip correctly

Cool, thanks.  Will now fill in the placeholders appropriately.  (The reason for all this fuss is to give the reader a bit of "why" behind the "do this" instructions, part of the effort to preempt some support calls.)

File: creating_images/index.adoc [hunk 11]
```diff
@@ -6,3 +6,7 @@
 :experimental:
 
 This guide provides best practices on writing and testing Docker images that can be used on OpenShift.
+ifdef::openshift-enterprise,openshift-origin[]
```
Comment (pweil-) [inline creating_images/index.adoc]: why is this scoped to just enterprise and origin?

**Comment by pweil-** [comment]
+1 to just restarting master to pick up the new IP if it was unable to be reused.  Thanks for putting this together.  Will take another look when the new changes land.

File: install_config/install/docker_registry.adoc (hunk 1 above)
Comment (bfallonf) [inline install_config/install/docker_registry.adoc]: Looks good!

File: creating_images/index.adoc (hunk 11 above)
Comment (tnguyen-rh) [inline creating_images/index.adoc]: @pweil-
The conditional is around a link to the Installation and Configuration document, which is OSO and OSE only, IIUC its [`_build_cfg.yml` Distros key](https://github.com/openshift/openshift-docs/blob/master/_build_cfg.yml#L152) correctly.
Am i missing something?

File: install_config/install/docker_registry.adoc [hunk 12]
```diff
@@ -169,9 +170,68 @@ $ oadm registry --service-account=registry \
...
+Re-using the Address::
+
+To re-use the IP address, you must save the IP address of the old docker-registry
+service prior to deleting it, and arrange to replace the newly assigned IP address
+with the saved one in the new docker-registry service.
+
+// NB: Snarfed from <https://github.com/openshift/openshift-docs/issues/1494>.
+. Make a note of the `ClusterIP` for the service.
+
+. Delete the service.
```
Comment (miminar) [inline install_config/install/docker_registry.adoc]: Save the address first with something like:

    oc describe svc/docker-registry | sed -n 's/^IP:\s*\(.*\)/\1/p'
    # or
    oc get svc/docker-registry -o yaml | grep 'clusterIP:'

?

Comment (tnguyen-rh) [inline install_config/install/docker_registry.adoc]: @miminar 
Thanks for the tip.  I think the second command is easier to understand on quick reading, and have added it.

File: creating_images/index.adoc (hunk 11 above)
Comment (pweil-) [inline creating_images/index.adoc]: Probably not, I was just trying to think of a reason that this link wouldn't be shown.  If install is keyed to those distros then this is fine.

**Comment by tnguyen-rh** [comment]
@pweil- @ncdc 
PTAL.

**Comment by pweil-** [comment]
Changes LGTM

**Comment by tnguyen-rh** [comment]
@ahardin-rh @adellape @bfallonf 
PTAL (2nd time for @bfallonf  due to big changes).
In particular, i would like to confirm that the labels are correct.

File: install_config/install/docker_registry.adoc [hunk 13]
```diff
@@ -169,9 +170,68 @@ $ oadm registry --service-account=registry \
 ----
 endif::[]
 
-[[viewing-logs]]
 
+[[registry-ip-address]]
+=== Registry IP Address
```
Comment (bfallonf) [inline install_config/install/docker_registry.adoc]: Can this have a more descriptive title? Something like "Defining the Registry IP Address" maybe? Something that describes the process the reader is about to undertake.

File: install_config/install/docker_registry.adoc [hunk 14]
```diff
@@ -169,9 +170,68 @@ $ oadm registry --service-account=registry \
...
+
+OpenShift refers to the integrated registry by its service IP address,
+so if you decide to delete and recreate the docker-registry service,
```
Comment (bfallonf) [inline install_config/install/docker_registry.adoc]: I think the markup for services is iffy in the rest of the docs, but should it have some *'s around docker-registry?

File: install_config/install/docker_registry.adoc [hunk 15]
```diff
@@ -169,9 +170,68 @@ $ oadm registry --service-account=registry \
...
+you can ensure a completely transparent transition by arranging to
+re-use the old IP address in the new service.
+If a new IP address cannot be avoided, you can minimize cluster
+disruption by rebooting only the masters.
+
+[[re-using-the-address]]
+Re-using the Address::
+
+To re-use the IP address, you must save the IP address of the old docker-registry
```
Comment (bfallonf) [inline install_config/install/docker_registry.adoc]: *'s here as well?

File: install_config/install/docker_registry.adoc [hunk 16]
```diff
@@ -169,9 +170,68 @@ $ oadm registry --service-account=registry \
...
+service prior to deleting it, and arrange to replace the newly assigned IP address
+with the saved one in the new docker-registry service.
+
+// NB: Snarfed from <https://github.com/openshift/openshift-docs/issues/1494>.
+. Make a note of the `ClusterIP` for the service.
```
Comment (bfallonf) [inline install_config/install/docker_registry.adoc]: Some :'s after the step descriptions leading into the command:

"Make a note of the `ClusterIP` for the service:"

File: install_config/install/docker_registry.adoc [hunk 17]
```diff
@@ -169,9 +170,68 @@ $ oadm registry --service-account=registry \
...
+----
+$ oc create -f registry.yaml
+----
+
+[[rebooting-the-masters]]
+Rebooting the Masters::
+
+If you are unable to re-use the IP address, any operation that uses a link:../../architecture/core_concepts/builds_and_image_streams.html#referencing-images-in-image-streams[pull specification]
+that includes the old IP address will fail.
+To minimize cluster disruption, you must reboot the masters.
```
Comment (bfallonf) [inline install_config/install/docker_registry.adoc]: Reboot command example?

----
# systemctl enable atomic-openshift-master
----

File: install_config/install/docker_registry.adoc (hunk 3 above)
Comment (bfallonf) [inline install_config/install/docker_registry.adoc]: @tnguyen-rh I think maybe I was misunderstood. This is a typo-level suggestion :P There's a word missing and I thought it'd be easier to make the change while you're in here already.

File: install_config/install/docker_registry.adoc [hunk 18]
```diff
@@ -390,6 +450,8 @@ In the following examples, we use:
 
 |====
 
+We omit the optional trailing `:<tag>`, so `<tag>` defaults to `latest`.
```
Comment (bfallonf) [inline install_config/install/docker_registry.adoc]: I'm not sure what this means. Who is we? Where does the <tag> go?

**Comment by bfallonf** [comment]
@tnguyen-rh A few comments, but cool :cool:

File: install_config/install/docker_registry.adoc (hunk 14 above)
Comment (adellape) [inline install_config/install/docker_registry.adoc]: I agree w/ the asterisk suggestion, at least as far as matching what we (mostly) use elsewhere for "unique names of objects".

Side note: I wish we used backticks more liberally in our markup guidelines, like for values of parameters/settings, and possibly for general "names of objects" like these. :disappointed:

File: install_config/install/docker_registry.adoc (hunk 13 above)
Comment (tnguyen-rh) [inline install_config/install/docker_registry.adoc]: @bfallonf 
> Can this have a more descriptive title? Something like "Defining the Registry IP Address" maybe? Something that describes the process the reader is about to undertake.
IIUC the procedure is only necessary if you want to remove and then re-add the `docker-registry` service.  The purpose of the procedure is to maintain the IP address, so perhaps a better title might be "Maintaining the Registry IP Address".

**Comment by tnguyen-rh** [comment]
@bfallonf @adellape 
Thanks for the review.  I've munged things according to your suggestions.  WDYT?

File: install_config/install/docker_registry.adoc (hunk 13 above)
Comment (bfallonf) [inline install_config/install/docker_registry.adoc]: @tnguyen-rh I like it!

File: install_config/install/docker_registry.adoc (hunk 14 above)
Comment (bfallonf) [inline install_config/install/docker_registry.adoc]: +1!

**Comment by bfallonf** [comment]
@tnguyen-rh Works for me :clap:

**Comment by tnguyen-rh** [comment]
**revhistory note:**

(Install & Config -> Install -> Deploying a Docker Registry)

- Mentioned default tag `latest`.
- Clarified importance of *<project-name>* in pull spec.
- Added section **Maintaining the Registry IP Address**.
